internal/controller/stages/regular_stages.go

```diff
@@ -0,0 +1,1981 @@
+				stage.Name,
+			),
+		},
+	); err != nil {
+		return fmt.Errorf(
+			"error listing Freight verified in Stage %q in namespace %q: %w",
+			stage.Name,
+			stage.Namespace,
+			err,
+		)
+	}
+
+	var errs []error
+	for _, f := range verified.Items {
+		newStatus := *f.Status.DeepCopy()
+		if newStatus.VerifiedIn == nil {
+			continue
+		}
+		delete(newStatus.VerifiedIn, stage.Name)
+
+		if err := kubeclient.PatchStatus(ctx, r.client, &f, func(status *kargoapi.FreightStatus) {
+			*status = newStatus
+		}); client.IgnoreNotFound(err) != nil {
+			errs = append(errs, fmt.Errorf(
+				"error clearing verification status of Freight %q in namespace %q: %w",
+				f.Name, f.Namespace, err,
+			))
+		}
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+// clearApprovals clears the approval status of all Freight that have been
+// approved for the given Stage. It removes the Stage from the ApprovedFor map
+// of each Freight.
+func (r *RegularStagesReconciler) clearApprovals(ctx context.Context, stage *kargoapi.Stage) error {
+	approved := kargoapi.FreightList{}
+	if err := r.client.List(
+		ctx,
+		&approved,
+		client.InNamespace(stage.Namespace),
+		client.MatchingFieldsSelector{
+			Selector: fields.OneTermEqualSelector(
+				indexer.FreightApprovedForStagesField,
+				stage.Name,
+			),
+		},
+	); err != nil {
+		return fmt.Errorf("error listing Freight approved for Stage %q in namespace %q: %w",
+			stage.Name,
+			stage.Namespace,
+			err,
+		)
+	}
+
+	var errs []error
+	for _, f := range approved.Items {
+		newStatus := *f.Status.DeepCopy()
+		if newStatus.ApprovedFor == nil {
+			continue
+		}
+		delete(newStatus.ApprovedFor, stage.Name)
+
+		if err := kubeclient.PatchStatus(ctx, r.client, &f, func(status *kargoapi.FreightStatus) {
+			*status = newStatus
+		}); client.IgnoreNotFound(err) != nil {
+			errs = append(errs, fmt.Errorf(
+				"error clearing approval status of Freight %q in namespace %q: %w",
+				f.Name, f.Namespace, err,
+			))
+		}
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+// clearAnalysisRuns clears all AnalysisRuns that are associated with the given
+// Stage. This is only done if the Rollouts integration is enabled.
+func (r *RegularStagesReconciler) clearAnalysisRuns(ctx context.Context, stage *kargoapi.Stage) error {
+	if !r.cfg.RolloutsIntegrationEnabled {
+		return nil
+	}
+
+	if err := r.client.DeleteAllOf(
+		ctx,
+		&rolloutsapi.AnalysisRun{},
+		client.InNamespace(stage.Namespace),
+		client.MatchingLabels(map[string]string{
+			kargoapi.StageLabelKey: stage.Name,
+		}),
+	); err != nil {
+		return fmt.Errorf("error deleting AnalysisRuns for Stage %q in namespace %q: %w",
+			stage.Name,
+			stage.Namespace,
+			err,
+		)
+	}
+	return nil
+}
+
+// summarizeConditions summarizes the conditions of the given Stage. It sets the
+// Ready condition based on the Promoting, Healthy and Verified conditions.
+// If there is an error, the Ready condition is set to False until the error is
+// resolved.
```

Comment:
I'm wondering if "ready" is the ideal condition to summarize the other conditions here. We've used ready for Warehouses and Projects, and for those cases, I think it makes sense as it implies "functional."

For a stage, I feel as if it's not obvious what "ready" means. Ready for what?

I am wondering whether something like steady or stable might better convey that all other conditions are nominal.

Reply:
The reason I picked "Ready" is because it is kind of standardized, and recognized by certain tools when they automatically try to determine if a resource is "in motion" or not (and thus, if they should e.g. wait for it to reach a stable state).

In this context, I see "ready" as "not working on anything new or worth waiting on".

Reply:
Language-wise, I have difficulty interpreting "ready" as meaning "idle", _buuuut_ the point about it being a common convention understood by a lot of tooling is a good one and well-taken. Consider me sufficiently convinced.